debian/pick-flame.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH PICK-FLAME 1 "June 27, 2001"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
pick-flame \- part of electricsheep, picks frames to render
.SH SYNOPSIS
.B pick-flame
.RI [ options ]
.SH DESCRIPTION
This program is part of the
.B electricsheep(1)
package and should not be used stand-alone.
.B Electric Sheep
starts this program when needed and supplies it with the correct parameters.
.SH AUTHOR
This manual page was written by Sander Smeenk <ssmeenk@debian.org>,
for the Debian GNU/Linux system (but may be used by others).


debian/changelog
electricsheep (3.0.2-git20191005-1ppa1~hirsute) hirsute; urgency=low

  * Rebuild for Ubuntu/Hirsute (21.04)
  * need to add LDFLAGS+=-lpthread (Autoconf fails to detect that)
  * can also be built with libwxgtk3.0-gtk3-dev (name change in Ubuntu)
  * Note: requires flam 3.1.1

 -- Hermann Vosseler <deb@ichthyostega.de>  Fri, 23 Apr 2021 18:23:55 +0100

electricsheep (3.0.2-git20191005-1ppa2~focal) focal; urgency=low

  * Rebuild for Ubuntu/Focal (20.04 LTS)
  * can also be built with libwxgtk3.0-gtk3-dev (name change in Ubuntu)
  * Note: requires flam 3.1.1

 -- Hermann Vosseler <deb@ichthyostega.de>  Wed, 24 Jun 2020 23:01:17 +0000

electricsheep (3.0.2-git20191005-1+buster) Lumiera-buster-tools; urgency=low

  * Build for Debian/Buster(stable).
  * Note: requires a port of glee-dev (dropped from Debian)

 -- Hermann Vosseler <deb@ichthyostega.de>  Mon, 06 Jan 2020 04:47:23 +0100

electricsheep (3.0.2-git20191005-1ppa1~bionic2) bionic; urgency=low

  * Rebuild for Ubuntu/Bionic (18.04 LTS)
  * Note: requires flam 3.1.1

 -- Hermann Vosseler <deb@ichthyostega.de>  Thu, 24 Oct 2019 16:33:33 +0200

electricsheep (3.0.2-git20191005-1~stretch) Lumiera-stretch-tools; urgency=low

  * Include upstream fixes on top of release 3.0.4 (2019-10-05)
  * Linux Version bump partially applied upstream
  * there are no further changes to flam3 upstream, so latest version is OK
  * build against lib versions from the Linux distro where applicable

 -- Hermann Vosseler <deb@ichthyostega.de>  Mon, 21 Oct 2019 05:37:57 +0200

electricsheep (3.0.2-git20180325-1ppa1~bionic1) bionic; urgency=low

  * Rebuild for Ubuntu/Bionic (18.04 LTS)
  * Note: needs flam 3.1.1

 -- Hermann Vosseler <deb@ichthyostega.de>  Sat, 19 May 2018 02:55:09 +0200

electricsheep (3.0.2-git20180325-1~stretch) Lumiera-stretch-tools; urgency=low

  * New upstream release 3.0.2 (2018-03-25)
  * require newest flam3 release.
  * update to build agaist libjpeg9
  * bump Debian standards version
  * update the installed desktop files
  * fix some FFmpeg deprecation warnings
  * build against lib versions from the Linux distro

 -- Hermann Vosseler <deb@ichthyostega.de>  Thu, 17 May 2018 17:50:05 +0200

electricsheep (2.7~b33-svn20150718-2ppa1~trusty) trusty; urgency=low

  * Rebuild for Ubuntu trusty (14.04 LTS)

 -- Hermann Vosseler <deb@ichthyostega.de>  Mon, 10 Aug 2015 01:40:20 +0200

electricsheep (2.7~b33-svn20150718-2) Lumera-jessie-tools; urgency=low

  * fix package definition to also work on Ubuntu

 -- Hermann Vosseler <deb@ichthyostega.de>  Mon, 10 Aug 2015 01:32:23 +0200

electricsheep (2.7~b33-svn20150718-1) lumiera-jessie-tools; urgency=low

  * New upstream snapshot from 2015-07-18
  * Start over with a 'all default automatic' debian package based on CDBS
  * Build for use on Debian/Jessie

 -- Hermann Vosseler <deb@ichthyostega.de>  Sun, 09 Aug 2015 16:56:33 +0200

electricsheep (2.7~b12+svn20091224-1.1) unstable; urgency=low

  * Non-maintainer upload.
  * Patch from Ubuntu by Fabrice Coutadeur to fix FTBFS. (Closes: #621985)
    * fix_ftbfs_libav_0.7.patch: fix deprecated variable to fix a FTBFS with
      libav 0.7
    * fix_link_as-needed.patch: fix Makefile.am and Makefile.in to avoid a FTBFS
      when as-needed is used with undefined gtk functions (LP: #770773) by moving
      libs from LDFLAGS to LDADD

 -- Konstantinos Margaritis <markos@debian.org>  Fri, 03 Feb 2012 11:50:38 +0200

electricsheep (2.7~b12+svn20091224-1) unstable; urgency=low

  * New upstream snapshot (Closes: #562230)

 -- Roberto C. Sanchez <roberto@connexer.com>  Thu, 31 Dec 2009 14:59:07 -0500

electricsheep (2.7~b12+svn20090708-2) unstable; urgency=low

  * Remove contrib script (esheep-mplayer.sh) from previous Debian
    package. (Closes: #546297)
  * debian/control:
    + Add Homepage field
    + Add dependency on mplayer
  * electricsheep.desktop.gnome: Remove obsolete --zoom option
  * electricsheep.desktop: Remove obsolete --zoom option

 -- Roberto C. Sanchez <roberto@connexer.com>  Wed, 30 Dec 2009 21:41:53 -0500

electricsheep (2.7~b12+svn20090708-1) unstable; urgency=low

  * New upstream release. (Closes: #533652)
  * New maintainer.
  * Move to dehelper compatibility level 5.
  * Update Build-Depends to remove obsolete package. (Closes: #515464)
  * Remove bashism from debian/rules. (Closes: #535377)
  * Remove unneeded dpatch patches.
  * Bump Standards-Version to 3.8.3 (no changes).
  * Clean up *.desktop files.
  * Add empty watch file explaining upstream's "release" policy.
  * Update debian/copyright to reflect new year and new maintainer.
  * Clean up upstream's man page (properly escape hyphens).

 -- Roberto C. Sanchez <roberto@connexer.com>  Mon, 24 Aug 2009 16:29:53 -0400

electricsheep (2.6.8-9) unstable; urgency=low

  [ Yves-Alexis Perez ]
  * link electricsheep to the correct location. (closes: #467441)

  [ Alexis Sukrieh ]
  * Standard Revision: 3.7.3
  * Changed Apps to Applications in debian/menu (fixed a lintian warning).

 -- Alexis Sukrieh <sukria@debian.org>  Tue, 11 Mar 2008 17:30:22 +0100

electricsheep (2.6.8-8) unstable; urgency=low

  * Added a debconf-2.0 alternate to dependencies (closes: #448477)

 -- Alexis Sukrieh <sukria@debian.org>  Mon, 19 Nov 2007 16:34:53 +0100

electricsheep (2.6.8-7) unstable; urgency=low

  * Summer cleanup of the package, install .desktop files in the appropriate
    places.
    (closes: #432782)
  * Autotools cleanup.

 -- Alexis Sukrieh <sukria@debian.org>  Fri, 13 Jul 2007 15:50:13 +0200

electricsheep (2.6.8-6) unstable; urgency=low

  * Compliant with gnome-screensaver
    + provides a valid electricsheep.desktop file in
    `/usr/share/gnome-screensaver/themes'.
    (closes: #425297)

 -- Alexis Sukrieh <sukria@debian.org>  Wed, 30 May 2007 13:37:45 +0200

electricsheep (2.6.8-5) unstable; urgency=low

  * Removed the debian revision from the version string so sheeps can be
    downloaded from the server.
    (closes: #391990)


flame/pick-flame.c
/*
    flame - cosmic recursive fractal flames
    Copyright (C) 1992  Scott Draves <spot@cs.cmu.edu>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/


#include "rect.h"


#define argi(s,d)   ((ai = getenv(s)) ? atoi(ai) : (d))
#define args(s,d)   ((ai = getenv(s)) ? ai : (d))
#define argd(s,d)   ((ai = getenv(s)) ? atof(ai) : (d))



rotcp(control_point *cp, double by)
{
   double s, t;
   int i;
   for (i = 0; i < NXFORMS; i++) {
      double r[2][2];
      double T[2][2];
      double U[2][2];
      double dtheta = by * 2.0 * M_PI / 360.0;

      r[1][1] = r[0][0] = cos(dtheta);
      r[0][1] = sin(dtheta);
      r[1][0] = -r[0][1];
      T[0][0] = cp->xform[i].c[0][0];
      T[1][0] = cp->xform[i].c[1][0];
      T[0][1] = cp->xform[i].c[0][1];
      T[1][1] = cp->xform[i].c[1][1];
      mult_matrix(r, T, U);
      cp->xform[i].c[0][0] = U[0][0];
      cp->xform[i].c[1][0] = U[1][0];
      cp->xform[i].c[0][1] = U[0][1];
      cp->xform[i].c[1][1] = U[1][1];
   }
}

int
main(argc, argv)
   int argc;
   char **argv;
{
   int count = 0;
   char s[4000], *ai, fname[50];
   unsigned char *image;
   control_point templ;
   control_point cp_orig;
   control_point cp_rotated;
   char *ss;
   int c, i, nstrips = 1;
   double avg_pix, fraction_black, fraction_white;
   double avg_thresh = argd("avg", 20.0);
   double black_thresh = argd("black", 0.01);
   double white_limit = argd("white", 0.05);
   int total_frames = argi("frames", 100);

   frame_spec f;

   srandom(time(0));

   strcpy(fname, "/tmp/pick.ppm");

   /* init the template */
   templ.background[0] = 0.0;
   templ.background[1] = 0.0;
   templ.background[2] = 0.0;
   templ.center[0] = 0.0;
   templ.center[1] = 0.0;
   templ.pixels_per_unit = 64;
   templ.width = 128;
   templ.height = 128;
   templ.spatial_oversample = 1;
   templ.gamma = 2.0;
   templ.vibrancy = 1.0;
   templ.hue_rotation = 0.0;
   templ.contrast = 1.0;
   templ.brightness = 1.0;
   templ.spatial_filter_radius = 0.5;
   templ.zoom = 0.0;
   templ.sample_density = 1;
   templ.nbatches = 10;
   templ.white_level = 200;
   templ.cmap_inter = 0;

   cp_orig = templ;

   f.temporal_filter_radius = 0.0;
   f.cps = &cp_orig;
   f.ncps = 1;
   image = (unsigned char *) malloc(3 * cp_orig.width * cp_orig.height);

   /* pick random control point until one looks good enough */

   do {
      int strip = 0;
      control_point cp;
      double height = cp_orig.height / (double) cp_orig.pixels_per_unit;
      double center_base = cp_orig.center[1] - ((nstrips-1)/2.0)*height/nstrips;

      unsigned char *strip_start =
	image + cp_orig.height * strip * cp_orig.width * 3;
      cp_orig.center[1] = center_base + height * (double) strip / nstrips;

      f.time = (double) 0.0;

      random_control_point(&cp_orig, -1);

      render_rectangle(&f, strip_start, cp_orig.width, field_both, 3, 0);

      if (1) {
	 FILE *fp;
	 fp = fopen(fname, "w");
	 if (!fp) {
	   perror("image out");
	   exit(1);
	 }
	 fprintf(fp, "P6\n");
	 fprintf(fp, "%d %d\n255\n", cp_orig.width, cp_orig.height);
	 fwrite(image, 1, 3 * cp_orig.width * cp_orig.height, fp);
	 fclose(fp);
      }

      if (1) {
	int i, n, tot, totb, totw;
	n = 3 * cp_orig.width * cp_orig.height;
	tot = 0;
	totb = 0;
	totw = 0;
	for (i = 0; i < n; i++) {
	  tot += image[i];
	  if (0 == image[i]) totb++;
	  if (255 == image[i]) totw++;

	  // printf("%d ", image[i]);
	}

	avg_pix = (tot / (double)n);
	fraction_black = totb / (double)n;
	fraction_white = totw / (double)n;

	if (0)
	  printf("avg pix = %g %g %g %g\n", avg_pix,
		 fraction_black, fraction_white, (double)n);
      } else {
	avg_pix = avg_thresh + 1.0;
	fraction_black = black_thresh + 1.0;
	fraction_white = white_limit - 1.0;
      }

   } while ((avg_pix < avg_thresh ||
	     fraction_black < black_thresh ||
	     fraction_white > white_limit) &&
	    count++ < 50);



   if (getenv("template")) {
      FILE *template = fopen(getenv("template"), "r");

      if (0 == template) {
	fprintf(stderr, "failed to open %s\n", getenv("template"));
	exit(1);
      }

      i = 0;
      ss = s;
      do {
	 c = getc(template);
	 if (EOF == c)
	    goto done_reading_template;
	 s[i++] = c;
      } while (';' != c);
	 
    done_reading_template:
      fclose(template);
      parse_control_point(&ss, &templ);
   }

   /* ugh */
   cp_orig.gamma = templ.gamma;
   cp_orig.vibrancy = templ.vibrancy;
   cp_orig.brightness = templ.brightness;
   cp_orig.background[0] = templ.background[0];
   cp_orig.background[1] = templ.background[1];
   cp_orig.background[2] = templ.background[2];
   cp_orig.spatial_oversample = templ.spatial_oversample;
   cp_orig.spatial_filter_radius = templ.spatial_filter_radius;
   cp_orig.sample_density = templ.sample_density;
   cp_orig.width = templ.width;
   cp_orig.height = templ.height;
   cp_orig.pixels_per_unit = templ.pixels_per_unit;

   if ((cp_orig.width & 15) ||
       (cp_orig.height & 15)) {
     fprintf(stderr, "pick-flame warning: width and height should be 0 mod 16, %dx%d is not.\n", cp_orig.width, cp_orig.height);
   }

   for (i = 0; i < 5; i++) {
     cp_orig.time = i * total_frames / 4;
     print_control_point(stdout, &cp_orig, 0);
     rotcp(&cp_orig, 90);
   }

   exit(0);
}
